src/main/java/com/pusher/client/channel/impl/PrivateEncryptedChannelImpl.java

```diff
@@ -124,6 +128,72 @@ public void updateState(ChannelState state) {
         }
     }
 
+    @Override
+    public PusherEvent prepareEvent(String event, String message) {
+        try {
+
+            Map receivedMessage = GSON.fromJson(message, Map.class);
+            final String decryptedMessage = decryptMessage((String) receivedMessage.get("data"));
+            receivedMessage.replace("data", decryptedMessage);
+
+            return GSON.fromJson(
+                    GSON.toJson(receivedMessage), PusherEvent.class);
+
+        } catch (AuthenticityException e1) {
+
+            // retry once only.
+            disposeSecretBoxOpener();
+            authenticate();
+
+            try {
+                Map receivedMessage = GSON.fromJson(message, Map.class);
+                final String decryptedMessage = decryptMessage((String) receivedMessage.get("data"));
+                receivedMessage.replace("data", decryptedMessage);
+
+                return GSON.fromJson(
+                        GSON.toJson(receivedMessage), PusherEvent.class);
+            } catch (AuthenticityException e2) {
+                disposeSecretBoxOpener();
+                notifyListenersOfDecryptFailure(event);
+            }
+        }
+        
+        return null;
+    }
+
+    private void notifyListenersOfDecryptFailure(final String event) {
+        Set<SubscriptionEventListener> listeners = getInterestedListeners(event);
+        if (listeners != null) {
+            for (SubscriptionEventListener listener : listeners) {
+                ((PrivateEncryptedChannelEventListener)listener).onDecryptionFailure(
+                        new Exception("Failed to decrypt message"));
```

Comment:
Not sure what kind of exception we should be returning here...

Reply:
I'm not even sure if we should be passing an exception in a callback. A simple string describing the error would be sufficient.

Reply:
It might be nice to pass some more details of the event as well. Perhaps the name of the event at least. I'm not sure the customer will have much use for the ciphertext or anything else though.

Reply:
So if we made the method `.onDecryptionFailure(String event, String reason)` - you think that would be sufficient? People will understand it's a failure callback, so we don't need to give them a packaged Exception to handle.